### PR TITLE
test: Run feature_bip68_sequence.py with MiniWallet

### PR DESCRIPTION
**Pull request by miles170**
This PR enables one more of the non-wallet functional tests (feature_bip68_sequence.py) to be run even when no wallet is compiled in by using the MiniWallet instead, as proposed in #20078.

### CODE REVIEW

**Comment by miles170** [comment]
I have no idea why the CI is failing. The error message is "Unauthorized! Configured service account doesn't have access.".

File: test/functional/feature_bip68_sequence.py [hunk 1]
```diff
@@ -92,16 +91,12 @@ def run_test(self):
...
         # Create some unconfirmed inputs
-        new_addr = self.nodes[0].getnewaddress()
-        self.nodes[0].sendtoaddress(new_addr, 2) # send 2 BTC
-
-        utxos = self.nodes[0].listunspent(0, 0)
-        assert len(utxos) > 0
+        self.wallet.send_self_transfer(from_node=self.nodes[0])
 
-        utxo = utxos[0]
+        utxo = self.wallet.get_utxo()
```
Comment (MarcoFalke) [inline test/functional/feature_bip68_sequence.py:96]: This may not be unconfirmed. If you need the unconfirmed one from above you can fetch is with `["new_utxo"]`.

Unrelated, I wonder why this needs to be unconfirmed

Comment (miles170) [reply test/functional/feature_bip68_sequence.py:96]: Should I change the code to match the origin behavior, which is unconfirmed UTXO?

```diff
diff --git a/test/functional/feature_bip68_sequence.py b/test/functional/feature_bip68_sequence.py
index e53e177ce..e14867dc3 100755
--- a/test/functional/feature_bip68_sequence.py
+++ b/test/functional/feature_bip68_sequence.py
@@ -91,9 +91,7 @@ class BIP68Test(BitcoinTestFramework):
     # the first sequence bit is set.
     def test_disable_flag(self):
         # Create some unconfirmed inputs
-        self.wallet.send_self_transfer(from_node=self.nodes[0])
-
-        utxo = self.wallet.get_utxo()
+        utxo = self.wallet.send_self_transfer(from_node=self.nodes[0])["new_utxo"]
 
         tx1 = CTransaction()
         value = int((utxo["value"] - self.relayfee) * COIN)
```

**Comment by miles170** [comment]
Force-pushed with takeing suggestion https://github.com/bitcoin/bitcoin/pull/26657#discussion_r1043466273 and remove formatting fixes.

**Comment by miles170** [comment]
@MarcoFalke  Is calling `gettxout` would affect `UTXOS_IN_CACHE` in `interface_usdt_utxocache.py`?
See https://cirrus-ci.com/task/5276905402793984

If so I have to find another way to ignore spent utxos in Miniwallet rescan_utxos. Or simply increase the `UTXOS_IN_CACHE` in `interface_usdt_utxocache.py`?

```
 test  2022-12-09T05:23:40.698000Z TestFramework.node0 (DEBUG): Node stopped 
 test  2022-12-09T05:23:40.698000Z TestFramework (INFO): handle_utxocache_flush(): UTXOCacheFlush(duration=4414, mode=ALWAYS, size=151, memory=31072, for_prune=False) 
 test  2022-12-09T05:23:40.699000Z TestFramework (INFO): handle_utxocache_flush(): UTXOCacheFlush(duration=4493, mode=ALWAYS, size=0, memory=2080, for_prune=False) 
 test  2022-12-09T05:23:40.733000Z TestFramework (INFO): check that we don't expect additional flushes 
 test  2022-12-09T05:23:40.733000Z TestFramework (ERROR): Assertion failed 
                                   Traceback (most recent call last):
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 134, in main
                                       self.run_test()
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/interface_usdt_utxocache.py", line 151, in run_test
                                       self.test_flush()
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/interface_usdt_utxocache.py", line 379, in test_flush
                                       assert_equal(0, len(expected_flushes))
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/util.py", line 56, in assert_equal
                                       raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
                                   AssertionError: not(0 == 1)
```

**Comment by miles170** [comment]
interface_usdt_utxocache.py is failing due to [test: Ignore spent utxos in MiniWallet rescan_utxos](https://github.com/bitcoin/bitcoin/pull/26657/commits/182bea1098b9d683e8698de4dd0a015f2b482338), I'm working on it.

File: test/functional/test_framework/wallet.py [hunk 2]
```diff
@@ -124,7 +125,16 @@ def rescan_utxos(self):
         res = self._test_node.scantxoutset(action="start", scanobjects=[self.get_descriptor()])
         assert_equal(True, res['success'])
         for utxo in res['unspents']:
-            self._utxos.append(self._create_utxo(txid=utxo["txid"], vout=utxo["vout"], value=utxo["amount"], height=utxo["height"]))
+            # Ignore utxo inside mempool
```
Comment (MarcoFalke) [inline test/functional/test_framework/wallet.py:128]: Can you explain why? `scantxoutset` doesn't scan the mempool, does it?

Comment (miles170) [reply test/functional/test_framework/wallet.py:128]: Without that the test throw txn-mempool-conflict, and get_it cos indeed return the tx that is inside the mempool.

File: test/functional/feature_bip68_sequence.py [hunk 3]
```diff
@@ -214,19 +200,20 @@ def test_sequence_lock_confirmed_inputs(self):
...
             if (using_sequence_locks and not should_pass):
                 # This transaction should be rejected
-                assert_raises_rpc_error(-26, NOT_FINAL_ERROR, self.nodes[0].sendrawtransaction, rawtx)
+                assert_raises_rpc_error(-26, NOT_FINAL_ERROR, self.wallet.sendrawtransaction, from_node=self.nodes[0], tx_hex=rawtx)
             else:
                 # This raw transaction should be accepted
-                self.nodes[0].sendrawtransaction(rawtx)
-                utxos = self.nodes[0].listunspent()
+                self.wallet.sendrawtransaction(from_node=self.nodes[0], tx_hex=rawtx)
+                self.wallet.rescan_utxos()
```
Comment (MarcoFalke) [inline test/functional/feature_bip68_sequence.py:215]: Why is this needed? `wallet.sendrawtransaction` should already account and scan the tx

Comment (miles170) [reply test/functional/feature_bip68_sequence.py:215]: `utxos = self.wallet.get_utxos(include_immature_coinbase = False)` implies `mark_as_spent = True`, which means utxos is cleared.
`wallet.sendrawtransaction` only appends the tx to `self._utxos`, so we need `self.wallet.rescan_utxos()` to rescan unspent utxos.

**Comment by miles170** [comment]
Force-pushed with squash commits and replace [test: Ignore spent utxos in MiniWallet rescan_utxos](https://github.com/bitcoin/bitcoin/pull/26657/commits/182bea1098b9d683e8698de4dd0a015f2b482338) with [test: Add "include mempool" flag to MiniWallet rescan_utxos](https://github.com/bitcoin/bitcoin/pull/26657/commits/90b9ab511b6d6b074dee6cabdd20bd42262392be)

File: test/functional/test_framework/wallet.py [hunk 4]
```diff
@@ -319,6 +309,20 @@ def create_self_transfer(self, *, fee_rate=Decimal("0.003"), fee=Decimal("0"), u
 
         return {"txid": tx["txid"], "wtxid": tx["tx"].getwtxid(), "hex": tx["hex"], "tx": tx["tx"], "new_utxo": tx["new_utxos"][0]}
 
+    def signrawtransactionwithwallet(self, *, tx):
```
Comment (MarcoFalke) [inline test/functional/test_framework/wallet.py:327]: Any reason for a duplicate method when `sign_tx` already exists?

Comment (miles170) [reply test/functional/test_framework/wallet.py:327]: `sign_tx` only works for `MiniWalletMode.RAW_P2PK`, and the default wallet mode is `MiniWalletMode.ADDRESS_OP_TRUE`, if we change `tx.vin` size we need `signrawtransactionwithwallet` to sign raw transaction on mode `MiniWalletMode.RAW_OP_TRUE` or `MiniWalletMode.ADDRESS_OP_TRUE`.

Comment (MarcoFalke) [reply test/functional/test_framework/wallet.py:327]: I mean you could modify sign_tx instead of adding a new method?

**Comment by miles170** [comment]
Force-pushed with takeing suggestion https://github.com/bitcoin/bitcoin/pull/26657#discussion_r1045596497

File: test/functional/test_framework/wallet.py [hunk 5]
```diff
@@ -142,20 +142,29 @@ def scan_tx(self, tx):
...
-        # 65 bytes: high-R val (33 bytes) + low-S val (32 bytes)
-        # with the DER header/skeleton data of 6 bytes added, this leads to a target size of 71 bytes
-        der_sig = b''
-        while not len(der_sig) == 71:
-            der_sig = self._priv_key.sign_ecdsa(sighash)
-            if not fixed_length:
-                break
-        tx.vin[0].scriptSig = CScript([der_sig + bytes(bytearray([SIGHASH_ALL]))])
-        tx.rehash()
+        """Sign tx that has been created by MiniWallet in P2TR/RAWSCRIPT/P2PK mode"""
```
Comment (MarcoFalke) [inline test/functional/test_framework/wallet.py:154]: in 0c20a83fa658fc660b138382d907615b9a60b845: Any reason to duplicate the list of modes, given that it just repeats all modes?

File: test/functional/test_framework/wallet.py [hunk 6]
```diff
@@ -119,12 +119,14 @@ def _bulk_tx(self, tx, target_weight):
...
         return sum(u['value'] for u in self._utxos)
 
-    def rescan_utxos(self):
+    def rescan_utxos(self, *, include_mempool=True):
         """Drop all utxos and rescan the utxo set"""
         self._utxos = []
         res = self._test_node.scantxoutset(action="start", scanobjects=[self.get_descriptor()])
         assert_equal(True, res['success'])
         for utxo in res['unspents']:
+            if not include_mempool and not self._test_node.gettxout(txid=utxo["txid"], n=utxo["vout"]):
```
Comment (MarcoFalke) [inline test/functional/test_framework/wallet.py:128]: This just seems to exclude utxos if they are spent in mempool transactions. Shouldn't it also include utxos if they are created in mempool transactions?

File: test/functional/test_framework/wallet.py (hunk 5 above)
Comment (miles170) [reply test/functional/test_framework/wallet.py:154]: Should I remove the comment or replace it with "Sign tx that has been created by MiniWallet"?

File: test/functional/test_framework/wallet.py (hunk 6 above)
Comment (miles170) [reply test/functional/test_framework/wallet.py:128]: [test_sequence_lock_confirmed_inputs](https://github.com/bitcoin/bitcoin/blob/69f0f6bc5ec9177cde7fdd0e73dac470f5641f82/test/functional/feature_bip68_sequence.py#L146-L156) would choose a random number of utxos from get_utxos, and if I included the transaction that is inside the mempool, the txn-mempool-conflict would occur.

Comment (MarcoFalke) [reply test/functional/test_framework/wallet.py:128]: I mean that someone setting the bool on the method would expect that this includes utxos created in the mempool. Also, I fail to see how including utxos created in the mempool can cause a conflict, as long as you remove utxos spent in the mempool.

This was also the behavior previously without MiniWallet, so it seems better to keep it, no?

Might be possible by calling scan_tx on all mempool txs, assuming they are topologically sorted.

Comment (miles170) [reply test/functional/test_framework/wallet.py:128]: Or should I use `getrawtransaction` to scan mempool, but this would require `txindex=1`.

Comment (MarcoFalke) [reply test/functional/test_framework/wallet.py:128]: `getrawmempool` does not require txindex

Comment (miles170) [reply test/functional/test_framework/wallet.py:128]: `getrawmempool` returns only the `transactionid` but not the `tx.vin`, which is required to scan spent UTXOs.

Comment (MarcoFalke) [reply test/functional/test_framework/wallet.py:128]: `getrawmempool` can return the transaction hex-encoded. You can decode it, or you may be able to call `getmempoolentry` if you need it decoded.

Comment (miles170) [reply test/functional/test_framework/wallet.py:128]: I'll see what I can do tomorrow morning, thanks!

Comment (miles170) [reply test/functional/test_framework/wallet.py:128]: > `getrawmempool` can return the transaction hex-encoded. You can decode it, or you may be able to call `getmempoolentry` if you need it decoded.

Sorry to bother you, but I just checked the bitcoin reference with [getrawmempool](https://developer.bitcoin.org/reference/rpc/getrawmempool.html) and [getmempoolentry](https://developer.bitcoin.org/reference/rpc/getmempoolentry.html), and there is no way to get `tx.vin` with these two methods.

Comment (MarcoFalke) [reply test/functional/test_framework/wallet.py:128]: Right, but `getrawtransaction` can

**Comment by miles170** [comment]
Force-pushed with taking suggestions https://github.com/bitcoin/bitcoin/pull/26657#discussion_r1045793617 and https://github.com/bitcoin/bitcoin/pull/26657#discussion_r1045795796

File: test/functional/test_framework/wallet.py [hunk 7]
```diff
@@ -119,12 +119,19 @@ def _bulk_tx(self, tx, target_weight):
...
         self._utxos = []
         res = self._test_node.scantxoutset(action="start", scanobjects=[self.get_descriptor()])
         assert_equal(True, res['success'])
+        spent_txs = set()
+        if not include_mempool:
+            for txid in self._test_node.getrawmempool():
+                tx = self._test_node.decoderawtransaction(self._test_node.getrawtransaction(txid))
+                spent_txs.update([(i["txid"], i["vout"]) for i in tx["vin"]])
         for utxo in res['unspents']:
+            if not include_mempool and (utxo["txid"], utxo["vout"]) in spent_txs:
```
Comment (MarcoFalke) [inline test/functional/test_framework/wallet.py:133]: Again, this will only *exclude* spent utxos in the mempool, but never *include* created ones, no?

File: test/functional/test_framework/wallet.py [hunk 8]
```diff
@@ -119,12 +119,19 @@ def _bulk_tx(self, tx, target_weight):
...
-    def rescan_utxos(self):
+    def rescan_utxos(self, *, include_mempool=True):
         """Drop all utxos and rescan the utxo set"""
         self._utxos = []
         res = self._test_node.scantxoutset(action="start", scanobjects=[self.get_descriptor()])
         assert_equal(True, res['success'])
+        spent_txs = set()
+        if not include_mempool:
+            for txid in self._test_node.getrawmempool():
+                tx = self._test_node.decoderawtransaction(self._test_node.getrawtransaction(txid))
```
Comment (MarcoFalke) [inline test/functional/test_framework/wallet.py:130]: Does getrawtransaction have a verbose options to skip the decode call?

File: test/functional/test_framework/wallet.py (hunk 7 above)
Comment (miles170) [reply test/functional/test_framework/wallet.py:133]: Do you mean the utxo created by vout?

Comment (MarcoFalke) [reply test/functional/test_framework/wallet.py:133]: Yes, a transaction in the mempool may have outputs (vout) that pay to the miniwallet

Comment (miles170) [reply test/functional/test_framework/wallet.py:133]: Thank you for your patience, and I'm sorry for the misunderstanding.

Comment (MarcoFalke) [reply test/functional/test_framework/wallet.py:133]: > Or could I simply assume that all outputs (vout) are paid to the miniwallet?

No, only the ones that pay to the address of the miniwallet.

Comment (miles170) [reply test/functional/test_framework/wallet.py:133]: There are so many different types of `scriptPubKey`, do we have a RPC API that can check if a certain `vout` belongs to the miniwallet?  Or should I write "helper method" to do the work?

Comment (MarcoFalke) [reply test/functional/test_framework/wallet.py:133]: You can use `scan_tx`

File: test/functional/test_framework/wallet.py [hunk 9]
```diff
@@ -119,12 +119,21 @@ def _bulk_tx(self, tx, target_weight):
     def get_balance(self):
         return sum(u['value'] for u in self._utxos)
 
-    def rescan_utxos(self):
+    def rescan_utxos(self, *, include_mempool=True):
         """Drop all utxos and rescan the utxo set"""
         self._utxos = []
         res = self._test_node.scantxoutset(action="start", scanobjects=[self.get_descriptor()])
         assert_equal(True, res['success'])
+        spents = set()
```
Comment (MarcoFalke) [inline test/functional/test_framework/wallet.py:127]: No need for this set if you move mempool logic last

File: test/functional/test_framework/wallet.py [hunk 10]
```diff
@@ -119,12 +119,21 @@ def _bulk_tx(self, tx, target_weight):
...
+        for txid in self._test_node.getrawmempool():
+            tx = self._test_node.getrawtransaction(txid=txid, verbose=True)
+            if include_mempool:
+                self.scan_tx(tx)
```
Comment (MarcoFalke) [inline test/functional/test_framework/wallet.py:131]: You will need to sort the transactions topologically, see also `BlockAssembler::SortForBlock`

File: test/functional/test_framework/wallet.py [hunk 11]
```diff
@@ -119,12 +119,21 @@ def _bulk_tx(self, tx, target_weight):
...
+            else:
+                spents.update([(i["txid"], i["vout"]) for i in tx["vin"]])
```
Comment (MarcoFalke) [inline test/functional/test_framework/wallet.py:133]: This will consider the mempool if `!include_mempool`, which may be confusing

Comment (miles170) [reply test/functional/test_framework/wallet.py:133]: What's proper way that you might consider reasonable?

Comment (MarcoFalke) [reply test/functional/test_framework/wallet.py:133]: No, just remove the else branch

Comment (miles170) [reply test/functional/test_framework/wallet.py:133]: If I remove the else branch `test_sequence_lock_confirmed_inputs` will encounter `txn-mempool-conflict`.

Comment (MarcoFalke) [reply test/functional/test_framework/wallet.py:133]: Even if you move the mempool scan last? See https://github.com/bitcoin/bitcoin/pull/26657#discussion_r1049630603

Comment (miles170) [reply test/functional/test_framework/wallet.py:133]: You are so right..

**Comment by miles170** [comment]
Force-pushed with taking suggestions https://github.com/bitcoin/bitcoin/pull/26657#discussion_r1049630603 and https://github.com/bitcoin/bitcoin/pull/26657#discussion_r1049631127 and https://github.com/bitcoin/bitcoin/pull/26657#discussion_r1049631862

File: test/functional/feature_bip68_sequence.py [hunk 12]
```diff
@@ -247,11 +234,12 @@ def test_sequence_lock_unconfirmed_inputs(self):
         tx2.nVersion = 2
         tx2.vin = [CTxIn(COutPoint(tx1.sha256, 0), nSequence=0)]
         tx2.vout = [CTxOut(int(tx1.vout[0].nValue - self.relayfee * COIN), SCRIPT_W0_SH_OP_TRUE)]
-        tx2_raw = self.nodes[0].signrawtransactionwithwallet(tx2.serialize().hex())["hex"]
+        self.wallet.sign_tx(tx=tx2)
+        tx2_raw = tx2.serialize().hex()
         tx2 = tx_from_hex(tx2_raw)
```
Comment (achow101) [inline test/functional/feature_bip68_sequence.py:239]: In 2512e1a37df42e47a1d2ea63511ba72c2ca8a92f "test: Run feature_bip68_sequence.py with MiniWallet"

This round trip is unnecessary.

Comment (miles170) [reply test/functional/feature_bip68_sequence.py:239]: Do you mean `tx_from_hex` is unnecessary?

```diff
diff --git a/test/functional/feature_bip68_sequence.py b/test/functional/feature_bip68_sequence.py
index 28f5acedb..56f4222a9 100755
--- a/test/functional/feature_bip68_sequence.py
+++ b/test/functional/feature_bip68_sequence.py
@@ -236,7 +236,6 @@ class BIP68Test(BitcoinTestFramework):
         tx2.vout = [CTxOut(int(tx1.vout[0].nValue - self.relayfee * COIN), SCRIPT_W0_SH_OP_TRUE)]
         self.wallet.sign_tx(tx=tx2)
         tx2_raw = tx2.serialize().hex()
-        tx2 = tx_from_hex(tx2_raw)
         tx2.rehash()

         self.wallet.sendrawtransaction(from_node=self.nodes[0], tx_hex=tx2_raw)
```

**Comment by miles170** [comment]
Force-pushed with taking suggestions https://github.com/bitcoin/bitcoin/pull/26657#discussion_r1062820833

**Comment by achow101** [comment]
ACK 4159ccd03142899028019a7cb44ee4120e68505a

File: test/functional/feature_bip68_sequence.py (hunk 3 above)
Comment (MarcoFalke) [reply test/functional/feature_bip68_sequence.py:215]: Ah, ok. So alternatively the `wallet.sendrawtransaction` could be dropped/reverted to just `sendrawtransaction`, because it is not needed to scan the tx.

File: test/functional/feature_bip68_sequence.py [hunk 13]
```diff
@@ -415,22 +402,19 @@ def activateCSV(self):
...
     # Use self.nodes[1] to test that version 2 transactions are standard.
     def test_version2_relay(self):
-        inputs = [ ]
-        outputs = { self.nodes[1].getnewaddress() : 1.0 }
-        rawtx = self.nodes[1].createrawtransaction(inputs, outputs)
-        rawtxfund = self.nodes[1].fundrawtransaction(rawtx)['hex']
-        tx = tx_from_hex(rawtxfund)
+        mini_wallet = MiniWallet(self.nodes[1])
+        mini_wallet.rescan_utxos()
+        tx = mini_wallet.create_self_transfer()["tx"]
```
Comment (MarcoFalke) [inline test/functional/feature_bip68_sequence.py:415]: nit in the last commit (can be fixed in a follow-up):

* MiniWallet shouldn't care about the node, so a single instance should suffice per test.
* Maybe the send_self_transfer function can accept a `version` option?

**Review by maflcko — APPROVED** [review]
Very nice. Thanks!


review ACK 4159ccd03142899028019a7cb44ee4120e68505a 🤸

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 4159ccd03142899028019a7cb44ee4120e68505a 🤸
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUisvgv/YsGnHt+lU5rGQdwFN6eIUrKwK/0pjUAIaDIeOp/61fuwDrJc3MhNE+/j
3cAiUyvHTtwo3eWkvV9FdQ9V+gVNOoiBWG1yEJtZNoM/syt5uaMi4zUvs2CisODn
tlr0jDcMSBw0E0Dc/YEKkwWytysPER5kPQiCleuzUbWqYD/QVX3xUOXGf+a0rttX
DpuzTcUu5ojaZh4YMEBkgprStIADIdxE2WPb/0eXd1pmwD+TpvnjVA/uRICQAft5
pBcePzrFLuGqJCOdsg5Fw72JmXJh1RjuBnCiOGiFL/xN02QH4R5MY401RNaXZtYO
B8USAMS8mO8lJB45t7njTpp1KSgrOJN0RJyEbja6BTROzyi6JttFJVRx4vie2hea
jICFDu8hNAWrkvXDQXWf/KaUSLOaWITDtYLRdK4bfirNy1PxiFsUjBPkrn0KJgaG
xI2AaYpNQqV7Krprdm9sEJpO8lev89aL+TEwZv7i2DjZONJArSOEE/Rtmm0nbQEI
2t90dylY
=isq1
-----END PGP SIGNATURE-----
```


</details>

File: test/functional/feature_bip68_sequence.py (hunk 13 above)
Comment (miles170) [reply test/functional/feature_bip68_sequence.py:415]: May I create another PR to refactor those nits?

Comment (MarcoFalke) [reply test/functional/feature_bip68_sequence.py:415]: Yes, if you agree with them and if the changes are possible (haven't checked myself)
